cpp/src/arrow/types/struct.h

```diff
@@ -23,7 +23,137 @@
 #include <vector>
 
 #include "arrow/type.h"
+#include "arrow/types/primitive.h"
+#include "arrow/types/list.h"
 
-namespace arrow {}  // namespace arrow
+namespace arrow {
 
-#endif  // ARROW_TYPES_STRUCT_H
+class StructArray : public Array {
+ public:
+  StructArray(const TypePtr& type, int32_t length,
+      std::vector<ArrayPtr>& values,
+      int32_t null_count = 0,
+      std::shared_ptr<Buffer> null_bitmap = nullptr) :
+      Array(type, length, null_count, null_bitmap) {
+    type_ = type;
+    values_ = values;
+  }
+
+  virtual ~StructArray() {}
+
+  // Return a shared pointer in case the requestor desires to share ownership
+  // with this array.
+  const std::shared_ptr<Array>& values(int32_t pos) const {return values_.at(pos);}
+  const std::vector<ArrayPtr>& values() const {return values_;}
```

Comment:
This overload introduces semantic confusion. Instead, make the API

`struct_arr->field(0)`

and

`struct_arr->fields()`

respectively